### PR TITLE
Dockerize client/UI

### PR DESCRIPTION
**Pull request by daniellrgn**
Move service from github actions/pages to Docker containers.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ENV NODE_ENV production
+
+WORKDIR ${DIRPATH}
+
+COPY ui ${DIRPATH}/ui
+COPY client ${DIRPATH}/client
+
+WORKDIR ${DIRPATH}/ui
+
+RUN npm ci
```
Comment (ivan-c) [inline Dockerfile]: Nice, wasn't [aware of `ci`](https://docs.npmjs.com/cli/v8/commands/npm-ci/) (vs `install`)! 

However, one of `ci`'s other aliases would be more descriptive

```suggestion
RUN npm clean-install
```

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+RUN npm run build
+RUN mv build public
+
+WORKDIR ${DIRPATH}/client
+
+RUN npm ci
```
Comment (ivan-c) [inline Dockerfile]:
```suggestion
RUN npm clean-install
```

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,28 @@
+FROM node:lts-bullseye as build-deps
```
Comment (ivan-c) [inline Dockerfile]:
```suggestion
FROM node:8 as build-deps
```
It's helpful to document the required version of node, but including it in the Dockerfile makes it even more explicit (and tracked by git!)

Pinning to a specific version also makes it machine-readable, which allows automated tools (eg [Renovate](https://github.com/renovatebot/renovate), [Dependabot](https://github.com/dependabot)) to audit and even automatically upgrade dependencies

File: docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,18 @@
+---
+version: "3.4"
+services:
+  server:
+    container_name: vaxx.link.server
```
Comment (ivan-c) [inline docker-compose.yml]: I'd avoid using `container_name`, even though it makes it super convenient to find your container.

When `container_name` is used in a dc project, only a single container may exist on the entire system with that name. ie I wouldn't be able to set up a 2nd copy of this project on the same VM
```suggestion
```

File: docker-compose.yml [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+      - "127.0.0.1:${EXTERNAL_PORT:-9000}:8000"
+    env_file:
+      - .env
+  client:
+    container_name: vaxx.link.client
+    build: .
+    ports:
+      - "127.0.0.1:${EXTERNAL_PORT:-9001}:3000"
+    env_file:
+      - .env
```
Comment (ivan-c) [inline docker-compose.yml]: Both NodeJS and [docker-compose](https://docs.docker.com/compose/environment-variables/#the-env-file) insist on using a config file named `.env`. 

I'd suggest using another name, just to make it clear where a config value is coming from.

```suggestion
      - app.env
```

File: .env [hunk 6]
```diff
@@ -0,0 +1 @@
+REACT_APP_REAL_SERVER_BASE=https://smart-health-links-server.cirg.washington.edu/api
```
Comment (ivan-c) [inline .env]: I'd avoid tracking this docker-compose/NodeJS config file in git. 

This app doesn't have any "secret" config values, but it's easy to accidentally commit secrets when a config file is tracked in git.

**Comment by ivan-c** [comment]
I tried reproducing the issues setting the CIRG domain (https://smart-health-links-server.cirg.washington.edu), by replacing each instance of the domain with a unique value and seeing which value took precedence.

I found that `REACT_APP_REAL_SERVER_BASE` only needed to be set once, in one of the below files (in order of precedence):

```
.env
Dockerfile
ui/.env
ui/src/App.tsx
```

I also noticed there are still websocket requests to port 3000, but as I mentioned before, I don't think it's a blocker. 

Apologies, I realized when I suggested using `NODE_ENV=production`, it was only 1/2 the solution. If I understand correctly, you'll also need to serve the output of `npm run build` from something other than `npm` (eg `serve`, or `nginx`. More than happy to help out if that's something you want to pursue.

**Review by ivan-c — APPROVED** [review]
Looks good!! Had a few suggestions, but no blockers for a prod deploy


To be 100% sure there weren't any caching issues involved, I tested each change with the below oneliner:
```
dc down -v && dc build --no-cache client && dc up -d
```

File: Dockerfile (hunk 3 above)
Comment (daniellrgn) [reply Dockerfile]: Thanks, locked the image, but bumped it to 18 as 8 was giving me issues in the build.

File: docker-compose.yml (hunk 5 above)
Comment (daniellrgn) [reply docker-compose.yml]: Changed file name to the suggested `app.env` and added `default.app.env`

File: .env (hunk 6 above)
Comment (daniellrgn) [reply .env]: Added default.app.env file and gitignored app.env

File: docker-compose.yml (hunk 4 above)
Comment (daniellrgn) [reply docker-compose.yml]: Thanks, this is really good to know! Removed.

File: client/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+WORKDIR /opt/app
+
+COPY package*.json ./
+RUN npm clean-install
+
+COPY . .
+RUN npm run build
+
+FROM nginx as production
+EXPOSE 80
```
Comment (ivan-c) [inline client/Dockerfile:12]: The `nginx` base image we're extending _should_ have `EXPOSE` already defined for those:

```
$ docker inspect nginx | grep -i expose -A3
            "ExposedPorts": {
                "80/tcp": {}
            },
            "Tty": false,
--
            "ExposedPorts": {
                "80/tcp": {}
            },
            "Tty": false,
```



```suggestion
```

File: ui/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,17 @@
+FROM node:18 as build-deps
+
+EXPOSE 3000
+
+# ENV DIRPATH /opt/app
```
Comment (ivan-c) [inline ui/Dockerfile:5]:
```suggestion
```

**Review by ivan-c — APPROVED** [review]
Looks good! More nitpicks, but still ready to merge!

Sorry for the late review again! :sweat_smile:

**Pull request merged**